Inital pe support

learn-more opened the pull request:
This PR is very WIP, but I am creating it anyway to gather some early feedback on the approach taken.
appveyor test result: https://ci.appveyor.com/project/learn-more/bloaty

Changes (per commit):
- Group projects in solution folders,  include headers in project
- Add clang-format
- Add PE test files compiled with Visual Studio 2015
- Initial PE implementation
  Only sections/segments are supported for now,
  the structures are from LIEF

#220

learn-more commented on src/pe.cc:
```diff
@@ -0,0 +1,315 @@
+// Copyright 2021 Google Inc. All Rights Reserved.
```
I was unsure how to handle attribution here, is this good as-is, should this be my name or something else?

learn-more commented on src/pe_structures.h:
```diff
@@ -0,0 +1,517 @@
+ * limitations under the License.
+ */
+
+// This is https://github.com/lief-project/LIEF/blob/master/include/LIEF/PE/structures.inc
```
This file is taken almost as-is from the project indicated.
I chose the LIEF headers since they also use the apache license, and I figured that would be compatible enough.

learn-more commented on tests/bloaty_test.cc:
```diff
@@ -15,7 +15,10 @@
 
 TEST_F(BloatyTest, EmptyObjectFile) {
-  std::string file = "01-empty.o";
+  if (pe_test_) {
+    GTEST_SKIP_("Object file not yet supported for PE");
+  }
+  std::string file = "01-empty." + ObjExtension();
```
I added some minimal changes to existing testcases (and skipped some that are not yet implemented).
Is this a good approach, or would you prefer a completely new test file with PE-specific tests?

haberman replied:
It's a great question. I've been thinking a lot about Bloaty's tests recently, see: https://github.com/google/bloaty/issues/221

The existing tests are not great: they aren't very robust or reproducible. The problem is that compilers have a lot of discretion when they create object files or binaries. That means that using input like C, Go, etc.doesn't constrain the output enough to have precise tests. We can see that problem here in your PR, where you had to loosen the precision of the tests to get the PE tests to pass. That unfortunately weakens the tests further.

I'm open to input here on the best way to proceed.

learn-more replied:
My suggestion to proceed would be:
- Move all current PE unittests to a new `bloaty_test_pe.cc`
- Here I will create new tests purely for the PE format
- Any overlap between this test and PE files with dwarf info will be something for a later date, when adding PE test files compiled with mingw. This could probably be joined with the current bloaty_test.cc, but we will cross that bridge when we encounter it.

learn-more commented on tests/test.h:
```diff
@@ -71,6 +71,13 @@ inline std::string DebugString(const google::protobuf::Message& message) {
  protected:
+
+  void SetUp() {
+    std::string file = "00-bloaty-pe";
+    uint64_t size;
+    pe_test_ = GetFileSize(file, &size);
```
Simple placeholder file used to detect PE or non-PE tests.

learn-more commented on tests/testdata/02-simple.c:
```diff
@@ -0,0 +1,10 @@
+
```
These source files are the ones used when building the samples, they are 1:1 matches from the ones generated in the .sh script.

learn-more commented on tests/testdata/make_test_files.bat:
```diff
@@ -0,0 +1,47 @@
+@echo off
```
Very crude script that builds msvc samples in-place (instead of in a temp folder like the .sh script)

learn-more commented on tests/test.h:
```diff
@@ -285,8 +292,24 @@ class BloatyTest : public ::testing::Test {
   }
 
+
+  const std::string ObjExtension() const {
+    if (pe_test_) {
+      return "obj";
+    }
+    return "o";
+  }
+
+  const std::string SharedObjectExtension() const {
+    if (pe_test_) {
+      return "dll";
+    }
+    return "so";
+  }
```
For the ones where linux used native extensions, I used windows extensions.
I also used .bin on the windows executable, since a lot of windows git configurations explicitly exclude *.exe

learn-more reviewed (commented):
@haberman I left some comments on my code to indicate choices I made,
please give some feedback on the approach taken, and how you want to move forward:
- Keep building in my repository until PE support is (somewhat) fleshed out
- or a more iterative approach, when a new feature is added create a new PR.

learn-more reviewed (commented):
I'll address the indicated issues, and update this PR when it's done.

If you dont' mind, I would like to force-push (rewriting the current commits), as to keep the history clean.

learn-more commented:
Updated the PR:

- Fixed formatting in pe.cc
- Moved LIEF header to third_party/lief_pe
- Added `third_party/lief_pe/METADATA`
- Added testfiles from msvc 2017 and msvc2019
- Used the test structure suggested in https://github.com/google/bloaty/issues/221#issuecomment-812950075
- Started a new test, that is more detailed.

Some questions:

1. Is the new test format good enough like this? Do you want to see more details or less details in the PE test?
2. This is the current PE output with `-v`:
```
Running bloaty: bloaty -v msvc-15.0-foo-bar.dll
FILE MAP:
0000-0290               656             [PE Headers]
0290-0400               368             [Unmapped]
0400-1200              3584             .text
1200-1a00              2048             .rdata
1a00-1c00               512             .data
1c00-1e00               512             .reloc

VM MAP:
0000-0290               656             [PE Headers]
0290-1000              3440             [-- Nothing mapped --]
1000-1c64              3172             .text
1c64-2000               924             [-- Nothing mapped --]
2000-26ec              1772             .rdata
26ec-3000              2324             [-- Nothing mapped --]
3000-338c               908             .data
338c-4000              3188             [-- Nothing mapped --]
4000-4128               296             .reloc
```
For the space between the sections:
- Windows actually rounds the allocations up to page size, how should I report this from the backend, as-is (not rounded up), or rounded up?

haberman commented:
This sounds great! I will try to take a look this weekend.

haberman commented on src/pe.cc:
```diff
@@ -0,0 +1,320 @@
+//! Sizes in bytes of various things in the COFF format.
+namespace STRUCT_SIZES {
+enum {
+  Header16Size = 20,
```
Why define these at all if we can just use `sizeof(pe_section)`, etc?

learn-more replied:
For the static asserts below.

haberman replied:
What is the purpose of asserting that these structs are a certain size? Are you trying to catch bugs in `third_party/lief_pe/pe_structures.h`? I think we can presume that the structures defined there are correct without double-checking them here.

learn-more replied:
Since we took a header file out of context, I'd like to verify that struct alignment is still correct.
We are using a different build system, so compiler flags / options might influence this.

haberman reviewed (commented):
> Is the new test format good enough like this? Do you want to see more details or less details in the PE test?

Yes, I think this is fine. I'm not inclined to invest too much into the current testing setup because I think it won't be truly robust until we use yaml2obj or similar.

> Windows actually rounds the allocations up to page size, how should I report this from the backend, as-is (not rounded up), or rounded up?

Generally I prefer to only report overheads that are present in the file itself. In this case it looks like `.text` in the file is 3584 bytes while `.text` in memory is 3172 bytes, plus 924 bytes of padding. So I would be inclined to have us report 3584 bytes in VM size, not 4096.  `.bss` is of course an exception to this, as `.bss` only exists in memory, in that case I think we should report whatever the section header says its size is.

> Instead of hardcoding the section sizes, I was thinking of putting a text file next to a PE that describes the actual file, this could be plain text, xml, yaml or anything really, what do you think about this?

What about the `bloaty -v` output itself?

learn-more commented:
> > Instead of hardcoding the section sizes, I was thinking of putting a text file next to a PE that describes the actual file, this could be plain text, xml, yaml or anything really, what do you think about this?
> 
> What about the `bloaty -v` output itself?

The bloaty format is not really machine-parsable,
unless you mean to textually match the tool output against the 'expected' output entirely?

haberman commented:
I was thinking of matching output exactly for now. If the bloaty memory map output changes, we can adapt at that point.

I'm happy to merge this as soon as tests are passing, even if it's not perfect. I think we can make the testing more robust as we go.

haberman commented:
Maybe we can mark it experimental for now, until it's more fleshed out.

learn-more commented:
The first commit is new, zlib did not build with msvc.
The second commit (Group projects in solution folders) was updated to also group zlib projects.

The 'testfiles' commit was updated to also include .txt files that have bloaty output.
The 'Initial PE implementation' commit was updated to test PE files by checking their output.

To make this more robust, it would make sense to 'normalize' output files:
* Trim trailing whitespace
* Replace \r\n with \n
* Maybe worry about localization?

haberman commented:
Normalizing sounds good.

It looks like you used Bloaty's normal output. I was thinking of using the memory map output, eg:

```
FILE MAP:
0000-0290               656             [PE Headers]
0290-0400               368             [Unmapped]
0400-1200              3584             .text
1200-1a00              2048             .rdata
1a00-1c00               512             .data
1c00-1e00               512             .reloc

VM MAP:
0000-0290               656             [PE Headers]
0290-1000              3440             [-- Nothing mapped --]
1000-1c64              3172             .text
1c64-2000               924             [-- Nothing mapped --]
2000-26ec              1772             .rdata
26ec-3000              2324             [-- Nothing mapped --]
3000-338c               908             .data
338c-4000              3188             [-- Nothing mapped --]
4000-4128               296             .reloc
```

haberman commented:
By the way, I'm going to merge https://github.com/google/bloaty/pull/231 now, which factors some commonly-used functions into a `util.h` file. You should be able to use that instead of redefining `CheckedAdd()`, `CheckedMul()`, `THROW`, etc.

learn-more commented:
> By the way, I'm going to merge #231 now, which factors some commonly-used functions into a `util.h` file. You should be able to use that instead of redefining `CheckedAdd()`, `CheckedMul()`, `THROW`, etc.



> Normalizing sounds good.
> 
> It looks like you used Bloaty's normal output. I was thinking of using the memory map output, eg:
> 
> ```
> FILE MAP:
> 0000-0290               656             [PE Headers]
> 0290-0400               368             [Unmapped]
> 0400-1200              3584             .text
> 1200-1a00              2048             .rdata
> 1a00-1c00               512             .data
> 1c00-1e00               512             .reloc
> 
> VM MAP:
> 0000-0290               656             [PE Headers]
> 0290-1000              3440             [-- Nothing mapped --]
> 1000-1c64              3172             .text
> 1c64-2000               924             [-- Nothing mapped --]
> 2000-26ec              1772             .rdata
> 26ec-3000              2324             [-- Nothing mapped --]
> 3000-338c               908             .data
> 338c-4000              3188             [-- Nothing mapped --]
> 4000-4128               296             .reloc
> ```

The memory map output is print directly to stdout with printf, so including that in the test would mean re-implementing most of the logic from `DualMaps`.

haberman commented:
> The memory map output is print directly to stdout with printf, so including that in the test would mean re-implementing most of the logic from DualMaps.

Let's just go with what you have, we can incrementally improve it once it's in. I really appreciate the contribution and I don't want to ask you to jump through too many hoops to get it in.

learn-more commented on src/pe.cc:
```diff
@@ -0,0 +1,280 @@
+#include "util.h"
+
+using absl::string_view;
+
```
Removed all helper functions now in util.h

learn-more commented on src/pe.cc:
```diff
@@ -0,0 +1,280 @@
+// limitations under the License.
+
+#include "bloaty.h"
+#include "absl/strings/substitute.h"
+#include "util.h"
```
`"absl/strings/substitute.h"` should technically be included inside `"util.h"` since one of the macro's there requires it (WARN)

learn-more commented on src/pe.cc:
```diff
@@ -0,0 +1,280 @@
+static_assert(STRUCT_SIZES::RelocationSize == sizeof(pe_relocation), "Compiler options broke LIEF struct layout");
+static_assert(STRUCT_SIZES::BaseRelocationBlockSize ==
+                  sizeof(pe_base_relocation_block),
+              "Compiler options broke LIEF struct layout");
```
Added better static assert message

learn-more commented on src/pe.cc:
```diff
@@ -0,0 +1,280 @@
+  // The last-line fallback to make sure we cover the entire VM space.
+  assert(pe.IsOpen());
+
+  auto begin = pe.header_region().data() - sink->input_file().data().data();
```
Had to fix this line, since absl uses std::string_view with c++17,
previously this was:
`auto begin = pe.header_region().begin() - sink->input_file().data().begin()`

However msvc iterators asserted on this, complaining that they were not compatible.
Might be something that could also be used to absl string_view iterators?

learn-more commented on tests/bloaty_test_pe.cc:
```diff
@@ -0,0 +1,126 @@
+}
+
+// Strip all trailing whitespace (including \r)
+void Normalize(std::string& contents) {
```
Added this `Normalize` function, that is used on both files read, and the bloaty output.

learn-more commented on tests/testdata/PE/x64/msvc-15.0-foo-bar.dll.txt:
```diff
@@ -0,0 +1,10 @@
+    FILE SIZE        VM SIZE    
+ --------------  -------------- 
```
This file is 1:1 bloaty output on windows, I left that in to test the 'Normalizing'.

haberman commented:
This is all great. Maybe just one more thing before we merge: if we are going with bloaty output instead of a memory map, could we use `--tsv` instead of the human-readable output. The `--tsv` format is at least intended to be machine-parseable, and has stronger stability guarantees.

learn-more commented on src/pe.cc:
```diff
@@ -0,0 +1,280 @@
+
+namespace bloaty {
+namespace pe {
+const uint16_t dos_magic = 0x5A4D;  // MZ
```
This used to be `'ZM'` but clang gave a warning on it.

learn-more commented on tests/bloaty_test_pe.cc:
```diff
@@ -0,0 +1,127 @@
+
+  bloaty::OutputOptions output_options;
+  std::stringstream output_stream;
+  output_options.output_format = bloaty::OutputFormat::kTSV;
```
The console output is still printed in 'human readable' format, only the validation text is formatted as TSV.

haberman commented:
I am happy to merge this now, are you ready?

learn-more commented:
It's 00.24 here, I am done for the day ;)

haberman commented:
Thanks for all your work on this!
